File: src/main/java/model/MatrixPath.java
package model;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.Getter;

public class MatrixPath {

    @Getter
    private final List<MatrixPoint> points;

    private final Set<MatrixPoint> pointSet;

    public MatrixPath(List<MatrixPoint> points) {
        this.points = points;
        pointSet = new HashSet<>(points);
    }

    public void showPath() {
        String stringPath = points.stream()
                .map((p) -> String.format("(%d, %d)", p.getX(), p.getY()))
                .collect(Collectors.joining(","));
        System.out.println(stringPath);
    }

    public boolean contains(MatrixPoint matrixPoint) {
        return pointSet.contains(matrixPoint);
    }
}

File: src/main/java/model/MatrixPath.kt
package model

import java.util.HashSet
import java.util.stream.Collectors

data class MatrixPath(val points: List<MatrixPoint>) {
    private val pointSet: Set<MatrixPoint>

    init {
        pointSet = HashSet(points)
    }

    fun showPath() {
        val stringPath = points.stream()
            .map { (x, y) -> "($x, $y)" }
            .collect(Collectors.joining(","))
        println(stringPath)
    }

    operator fun contains(matrixPoint: MatrixPoint): Boolean {
        return pointSet.contains(matrixPoint)
    }
}
